saleor/graphql/product/mutations/channels.py

```diff
@@ -156,12 +171,103 @@ def add_channels(cls, product: "ProductModel", add_channels: List[Dict]):
-            ProductChannelListing.objects.update_or_create(
+            product_channel_listing, _ = ProductChannelListing.objects.update_or_create(
                 product=product, channel=channel, defaults=defaults
             )
+            cls.validate_variants(add_variants, remove_variants)
+            cls.add_variants(info, channel, add_variants)
+            cls.remove_variants(
+                info, product_channel_listing, product, channel, remove_variants
+            )
+
+    @classmethod
+    def validate_variants(cls, add_variants, remove_variants):
+        error_code = ProductErrorCode.DUPLICATED_INPUT_ITEM.value
+        errors = defaultdict(list)
+        cls.validate_duplicated_variant_ids(
+            add_variants, remove_variants, errors, error_code
+        )
+        if errors:
+            raise ValidationError(errors)
+
+    @classmethod
+    def validate_duplicated_variant_ids(
+        cls,
+        add_variant_ids: Iterable[str],
+        remove_variant_ids: Iterable[str],
+        errors: ErrorType,
+        error_code,
+    ):
+        duplicated_ids = get_duplicates_ids(add_variant_ids, remove_variant_ids)
+        if duplicated_ids:
+            error_msg = (
+                "The same object cannot be in both lists "
+                "for adding and removing items."
+            )
+            errors["input"].append(
+                ValidationError(
+                    error_msg,
+                    code=error_code,
+                    params={"variants": list(duplicated_ids)},
+                )
+            )
+
+    @classmethod
+    def add_variants(cls, info, channel, add_variants: List[Dict]):
+        if not add_variants:
+            return
+        variants_id = [
+            cls.get_node_or_error(info, id, only_type=ProductVariant, field="id")
+            for id in add_variants
+        ]
+
+        variant_channel_listings = []
+        for variant in variants_id:
+            variant_channel_listings.append(
+                ProductVariantChannelListing(channel=channel, variant=variant)
+            )
+
+        try:
+            ProductVariantChannelListing.objects.bulk_create(variant_channel_listings)
+        except IntegrityError:
+            raise ValidationError(
+                {
+                    "addVariants": ValidationError(
+                        "One of channel listing already "
+                        "exists for this product variant.",
+                        code=ProductErrorCode.ALREADY_EXISTS.value,
+                    )
+                }
+            )
+
+    @classmethod
+    def remove_variants(
+        cls,
+        info,
+        product_channel_listing,
+        product,
+        channel,
+        remove_variants: List[Dict],
+    ):
+        if not remove_variants:
+            return
+        variants_id = [
+            cls.get_node_or_error(info, id, only_type=ProductVariant, field="id")
+            for id in remove_variants
+        ]
```

Comment:
Firstly, `get_node_or_error` returns iinstance, not `id`, secondly, you can use `get_nodes_or_error` for getting all instances.

Reply:
Same situation in `add_variants` method.